### PR TITLE
feat(capture): add an SGLang main (923e4a56) target for the spec-capture patch

### PR DESCRIPTION
**Pull request by bighead-liat**
## Motivation

Server-side capture (`--enable-spec-capture`) is how every disaggregated recipe feeds a trainer, and the checked-in patches target SGLang `v0.5.18` (2026-08-20) and three Kimi K3 revisions. DeepSeek-V4.1, and anything newer than v0.5.18, only serves on SGLang main: main is 692 commits past v0.5.18, and `git apply --check` of the v0.5.18 patch against main at `923e4a56` rejects 16 hunks in 9 files. Until the patch moves, SpecForge cannot capture from a V4.1 target at all.

## Modifications

- `patches/sglang/main-923e4a56/spec-capture.patch`: the v0.5.18 patch ported to main at `923e4a56` (2026-09-16). Same sink (`spec_capture_sink.py`), same writer-rank and pending-batch logic, same capture-only DSpark routing. What had to move with main:
  - the three server args are declared in the `ExecFeatures` arg-group struct and `field_order.py` (the flat `ServerArgs` fields with `NS(...)` annotations are gone); `--enable-spec-capture`, `--spec-capture-aux-layer-ids` and `--spec-capture-method {eagle3,dflash,dspark}` are generated from them unchanged;
  - capture-only aux wiring writes into the runner's `spec_aux_config` (`SpecAuxHiddenStateConfig`) instead of the removed runner attributes;
  - the per-iteration drain of pending capture transfers lives in `Scheduler.on_idle`, ahead of the fully-idle gate, where the old idle path used to be;
  - `GenerationBatchResult.copy_to_cpu` gained the `capture_stream` snapshot path at its new call sites;
  - the capture branches in the result processor and streamer test the request with `getattr(req, "spec_capture", None)` because upstream unit tests drive them with request stubs.
- `scripts/apply_sglang_spec_capture_patch.sh`: `--target main-923e4a56` (no version prefix; a main checkout reports a dev version, so `patch --check` is the gate, as for the Kimi targets).
- `examples/configs/README.md`: when the main target is needed and how to regenerate it for a newer revision.
- Tests: `test_disagg_launchers.py` pins the ported patch's invariants (writer-rank checks, sink batching, capture-only DSpark routing, arg-group declarations, stub-safe guards); `test_apply_sglang_spec_capture_patch.py` covers the new target.

## Accuracy Test

SGLang main `923e4a56` with the patch applied, one RTX 4090:

- `tests/test_runtime/test_server_capture_gate.py` (`SPECFORGE_RUN_SERVER_CAPTURE_TESTS=1`): **2 passed** in 57 s. A live patched server writes into a live Mooncake store, `test_eagle3_zero_copy_end_to_end` matches the captured aux/last hidden states against an HF reference forward and runs one real train step, `test_dflash_capture_same_server` captures with the dflash schema on the same server.
- SGLang's own unit tests on the patched tree: `test/registered/unit/managers/` (result processor, streamer, aux output, sampling-mask), `test/registered/unit/server_args/` and `test_serving_chat.py` in one invocation: 510 passed, 1 failed. The failure (`test_server_args.py::TestMultimodalFeatureTransport::test_default_transport_is_cpu_for_unsupported_multinode_model`) is order-dependent and reproduces identically on pristine `923e4a56` with the same invocation; it passes in isolation on both trees. pre-commit clean.
- `git apply --check -p2` of the patch against a pristine `923e4a56` checkout from the package parent, exactly as the apply script does: clean.
- SpecForge: `tests.test_scripts.test_disagg_launchers` and `tests.test_scripts.test_apply_sglang_spec_capture_patch`: 17 tests, OK; pre-commit clean.

Not verified here: a DeepSeek-V4.1 target. The V4.1-specific hook (`DeepseekV4ForCausalLM.set_dspark_layers_to_capture`) is upstream and untouched by the patch; the model needs an 8-GPU node I do not have. I will report a V4.1 capture run once I do.

## Related

- The v0.5.18 and Kimi K3 targets are unchanged; this adds a target, it does not replace one.

## Benchmark & Profiling

Not applicable.

### CODE REVIEW

File: patches/sglang/main-923e4a56/spec-capture.patch [hunk 1]
```diff
@@ -0,0 +1,1342 @@
...
++        # while SpecForge folds the same multiplier into its frozen target
++        # head. Restore the pre-head-scale representation before capture so
++        # the multiplier is applied exactly once when training recomputes logits.
++        logits_mup_width_multiplier = getattr(
++            self.config, "logits_mup_width_multiplier", None
++        )
++        if last_hidden_states_to_store is not None and logits_mup_width_multiplier:
++            last_hidden_states_to_store = last_hidden_states_to_store * float(
++                logits_mup_width_multiplier
++            )
```
[the remark this answers is not in the document]
Comment (bighead-liat) [reply patches/sglang/main-923e4a56/spec-capture.patch:103]: The multiply is the right direction, and it is verbatim from the v0.5.18 patch (the inventory documents it). The chain on main: Inkling divides the post-norm hidden by the multiplier before it reaches `LogitsProcessor` (`models/inkling.py`: `hidden_states_for_logits = hidden_states / mup_width_multiplier`), so the tensor `LogitsProcessor` sees is the scaled-down one. The capture restores the pre-division representation by multiplying back. On the training side SpecForge folds the multiplier into the frozen head by dividing its weights (`target_utils.py`: `lm_head.weight.data.div_(mup_multiplier)`), so training computes `(W / m)(h) = W(h / m)`, the serving logits, exactly once. SGLang's own DSpark draft says the same thing from the other side (`models/dspark.py`: trained against a folded head, serving divides once). Dividing here instead would apply the factor twice.

File: patches/sglang/main-923e4a56/spec-capture.patch [hunk 2]
```diff
@@ -0,0 +1,1342 @@
...
++            tensor.device
++            for _, aux, last_hidden in samples
++            for tensor in (aux, last_hidden)
++            if tensor is not None and tensor.is_cuda
++        }
++        ready = []
++        for device in devices:
++            event = torch.cuda.Event()
++            event.record(torch.cuda.current_stream(device))
++            ready.append(event)
```
[the remark this answers is not in the document]
Comment (bighead-liat) [reply patches/sglang/main-923e4a56/spec-capture.patch:1075]: The event is recorded after the snapshot has already completed. In the overlap path the clone runs on the copy stream inside `copy_to_cpu`, `copy_done` is recorded on that stream right after, and the prefill result processor calls `result.copy_done.synchronize()` (`batch_result_processor.py`, before any request is staged) ahead of `_sink_spec_capture` / `_queue_spec_captures`, which is where `submit_samples` records its event. By then the clone is finished on the host side, so which stream is current at that point does not matter. This is also unchanged from v0.5.18, and the GPU put path only switches on with `MOONCAKE_PROTOCOL=rdma`; the gate run in the description used the host path.

File: examples/configs/README.md [hunk 3]
```diff
@@ -89,6 +89,15 @@ servers. Its
 covers the v0.5.18 SGLang capture patch and the bundled `deepseek-v4` chat
 template (the checkpoint ships no Jinja template).
 
+Targets that only run on SGLang main (DeepSeek-V4.1 and anything newer than
+v0.5.18) use the same capture patch ported to main:
+`scripts/apply_sglang_spec_capture_patch.sh --target main-923e4a56` applies
+`patches/sglang/main-923e4a56/spec-capture.patch` to a checkout or install
+of SGLang main at 923e4a56 (2026-09-16). It exposes the same
+`--enable-spec-capture` / `--spec-capture-method` / `--spec-capture-aux-layer-ids`
+flags, now declared in SGLang's arg-group structs. Re-generate the patch
```
[the remark this answers is not in the document]
Comment (bighead-liat) [reply examples/configs/README.md:98]: Done in the follow-up commit: `specforge/inference/sglang_patch_inventory.md` has the target row and the apply line, and the `server_capture` adapter docstring now points at the inventory instead of naming v0.5.18.

**Comment by bighead-liat** [comment]
Replies to the three suppressed comments in the Copilot review:

**Omitted `--spec-capture-aux-layer-ids`.** Correct as a description of the current behaviour, and it is the same on v0.5.18: the sink records `aux_layer_ids: None` and the SpecForge client refuses a capture whose requested ids cannot be verified. In the SpecForge flow the producer always sends explicit ids from the draft config, so the omission fails loudly at the client rather than silently training on default layers. I kept the port behaviour-identical to v0.5.18 here; resolving the model's effective default ids in the sink (or making the flag required) is a change for both targets and I can send it separately.

**Duplicate finished presentation.** The "presented twice" note in the streamer is about the decode side under the overlap schedule. Capture requests are staged from the prefill result path, and with chunked prefill disabled (the server refuses to start otherwise) a request belongs to exactly one prefill batch, so `_sink_spec_capture` runs once per request. The streamer's capture gate additionally holds a finished capture request until its result is set. Same logic as v0.5.18.

**Test coverage.** Agreed that string assertions are not runtime coverage. The follow-up commit adds `test_main_923e4a56_patch_applies_to_the_pinned_checkout`: with `SPECFORGE_SGLANG_MAIN_CHECKOUT` pointing at an sglang checkout it runs the same `git apply --check -p2` the script runs, from the package parent with repository discovery stopped there (it passes against `923e4a56`). It skips in CI, which has no main checkout; the runtime coverage for this target is the server-capture gate run recorded in the description. A CI job that checks out main is worth having but is a workflow change I would rather not fold into this PR.
